File: app/src/main/java/com/solt9029/editmasterandroid/view/EndlessScrollListener.java
package com.solt9029.editmasterandroid.view;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public abstract class EndlessScrollListener extends RecyclerView.OnScrollListener {
    private LinearLayoutManager manager;
    public static final int VISIBLE_THRESHOLD = 1;

    EndlessScrollListener(LinearLayoutManager manager) {
        this.manager = manager;
    }

    @Override
    public void onScrolled(RecyclerView view, int dx, int dy) {
        super.onScrolled(view, dx, dy);

        // prevent onScrolled from being called when error happens. onScrolled is called when item count changes!
        if (dy <= 0) {
            return;
        }

        int visibleItemCount = view.getChildCount();
        int totalItemCount = manager.getItemCount();
        int firstVisibleItem = manager.findFirstVisibleItemPosition();

        if ((totalItemCount - visibleItemCount) <= (firstVisibleItem + VISIBLE_THRESHOLD)) {
            onLoadMore();
        }
    }

    public abstract void onLoadMore();
}

File: app/src/main/java/com/solt9029/editmasterandroid/view/EndlessScrollListener.kt
package com.solt9029.editmasterandroid.view

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

abstract class EndlessScrollListener protected constructor(
        private val manager: LinearLayoutManager
) : RecyclerView.OnScrollListener() {
    override fun onScrolled(view: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(view, dx, dy)

        // prevent onScrolled from being called when error happens. onScrolled is called when item count changes!
        if (dy <= 0) {
            return
        }

        val visibleItemCount = view.childCount
        val totalItemCount = manager.itemCount
        val firstVisibleItem = manager.findFirstVisibleItemPosition()

        if (totalItemCount - visibleItemCount <= firstVisibleItem + VISIBLE_THRESHOLD) {
            onLoadMore()
        }
    }

    abstract fun onLoadMore()

    companion object {
        private const val VISIBLE_THRESHOLD = 1
    }
}